client/telemetry/src/lib.rs

```diff
@@ -16,339 +16,386 @@
-pub fn init_telemetry(config: TelemetryConfig) -> Telemetry {
-	// Build the list of telemetry endpoints.
-	let (endpoints, wasm_external_transport) = (config.endpoints.0, config.wasm_external_transport);
-
-	let (sender, receiver) = mpsc::channel(16);
-	let guard = {
-		let logger = TelemetryDrain { sender: std::panic::AssertUnwindSafe(sender) };
-		let root = slog::Logger::root(slog::Drain::fuse(logger), slog::o!());
-		slog_scope::set_global_logger(root)
-	};
-
-	let worker = match worker::TelemetryWorker::new(endpoints, wasm_external_transport) {
-		Ok(w) => Some(w),
-		Err(err) => {
-			error!(target: "telemetry", "Failed to initialize telemetry worker: {:?}", err);
-			None
+	/// Run the telemetry worker.
+	///
+	/// This should be run in a background task.
+	pub async fn run(self) {
+		let Self {
+			receiver,
+			sender: _sender,
+			mut init_receiver,
+			init_sender: _init_sender,
+			transport,
+		} = self;
+
+		let mut node_map: HashMap<Id, Vec<(u8, Multiaddr)>> = HashMap::new();
+		let mut node_args: HashMap<
+			Multiaddr,
+			(Vec<ConnectionMessage>, Vec<ConnectionNotifierSender>),
+		> = HashMap::new();
+		let mut existing_nodes: HashSet<Multiaddr> = HashSet::new();
+
+		// initialize the telemetry nodes
+		init_receiver.close();
+		while let Some((id, endpoints, connection_message, mut connection_notifiers_rx)) =
+			init_receiver.next().await
+		{
+			let endpoints = endpoints.0;
+
+			connection_notifiers_rx.close();
+			let connection_notifier_senders: Vec<_> = connection_notifiers_rx.collect().await;
+
+			for (addr, verbosity) in endpoints {
+				node_map
+					.entry(id.clone())
+					.or_insert_with(Vec::new)
+					.push((verbosity, addr.clone()));
+				existing_nodes.insert(addr.clone());
+				let mut obj = connection_message.clone();
+				obj.insert("msg".into(), "system.connected".into());
+				obj.insert("id".into(), id.into_u64().into());
+				let (connection_messages, connection_notifiers) = node_args
+					.entry(addr.clone())
+					.or_insert_with(|| (Vec::new(), Vec::new()));
+				connection_messages.push(obj);
+				connection_notifiers.extend(connection_notifier_senders.clone());
+			}
 		}
-	};
 
-	Telemetry {
-		inner: Arc::new(Mutex::new(TelemetryInner {
-			worker,
-			receiver,
-		})),
-		_guard: Arc::new(guard),
-	}
-}
+		let mut node_pool: Dispatcher = existing_nodes
+			.iter()
+			.map(|addr| {
+				let (connection_messages, connection_notifiers) = node_args
+					.remove(addr)
+					.expect("there is a node for every connection message; qed");
+				let node = Node::new(
+					transport.clone(),
+					addr.clone(),
+					connection_messages,
+					connection_notifiers,
+				);
+				(addr.clone(), node)
+			})
+			.collect();
 
-/// Event generated when polling the worker.
-#[derive(Debug)]
-pub enum TelemetryEvent {
-	/// We have established a connection to one of the telemetry endpoint, either for the first
-	/// time or after having been disconnected earlier.
-	Connected,
+		let _ = receiver
```

Comment:
It didn't work with the async function. It was too complicated to handle.

Since we have a stream (telemetry events) and a sink (network connection), why not used them as stream and sink?

Reply:
Because this code is too complicated.

Reply:
I will try

Reply:
Ty

Reply:
Closing this duplicated comment in favor of https://github.com/paritytech/substrate/pull/7463#discussion_r553330059